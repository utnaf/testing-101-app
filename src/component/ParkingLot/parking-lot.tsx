import { Grid, Card, CardMedia, CardContent, Typography, CardActions, Button, Box, Fab } from '@material-ui/core';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { IRootState } from 'src/shared/reducer';
import { getParkingLots } from './parking-lot.reducer';
import Alert from '@material-ui/lab/Alert';
import AddIcon from '@material-ui/icons/Add';
import Loader from 'src/shared/components/Loader';
import { Link, RouteComponentProps } from 'react-router-dom';

import getClasses from './style';

export interface IParkingLotProps extends StateProps, DispatchProps, RouteComponentProps {}

export const ParkingLot = (props: IParkingLotProps) => {
  const { parkingLots, loading } = props;

  const classes = getClasses();

  useEffect(() => {
    props.getParkingLots();
  }, []);

  const openLotDashboard = id => e => {
    e.preventDefault();
    props.history.push(`/lot/${id}`);
  };

  return (
    <Box>
      <Loader visible={loading} />
      {parkingLots.length > 0 ? (
        <Grid container spacing={4}>
          {parkingLots.map(parkingLot => (
            <Grid item xs={12} sm={6} md={4} key={parkingLot.id} onClick={openLotDashboard(parkingLot.id)}>
              <Card className={classes.card}>
                <CardMedia className={classes.media} image={parkingLot.imageUrl} title={`${parkingLot.name} Picture`} />
                <CardContent>
                  <Grid container justify="space-between">
                    <Grid item xs={8}>
                      <Typography gutterBottom variant="h5" component="h2">
                        {parkingLot.name}
                      </Typography>
                    </Grid>
                    <Grid item>
                      <Typography gutterBottom variant="h5" component="small">
                        {parkingLot.status.carCount} / {parkingLot.spots}
                      </Typography>
                    </Grid>
                  </Grid>

                  <Typography>{parkingLot.address}</Typography>
                </CardContent>
              </Card>
            </Grid>
          ))}
        </Grid>
      ) : (
        <Alert severity="info">Non hai ancora nessun Parcheggio da gestire.</Alert>
      )}
      <Fab color="secondary" aria-label="add parking lot" className={classes.fab} component={Link} to="/lot/create">
        <AddIcon />
      </Fab>
    </Box>
  );
};

const mapStateToProps = ({ parkingLot }: IRootState) => ({
  parkingLots: parkingLot.entities,
  loading: parkingLot.loading,
});

const mapDispatchToProps = {
  getParkingLots,
};

type StateProps = ReturnType<typeof mapStateToProps>;
type DispatchProps = typeof mapDispatchToProps;

export default connect(mapStateToProps, mapDispatchToProps)(ParkingLot);
