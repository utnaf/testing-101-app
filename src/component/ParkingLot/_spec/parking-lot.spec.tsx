import React from 'react';
import { shallow } from 'enzyme';
import { ParkingLot } from '../parking-lot';
import { IParkingLot } from '../parking-lot.model';
import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';

const history = createMemoryHistory();
const path = `/lot`;
const fakeMatch: match = {
  isExact: false,
  path,
  url: path,
  params: {},
};
const location = createLocation(fakeMatch.url);

const getParkingLotsDummy = () => null;

const baseParkingLot = {
  id: 1,
  name: 'Foo',
  address: 'Bar',
  spots: 100,
  status: {
    carCount: 10,
  },
} as IParkingLot;

const baseProps = {
  parkingLots: [baseParkingLot],
  loading: false,
  getParkingLots: getParkingLotsDummy,
  location,
  match: fakeMatch,
  history,
};

let mountedWrapper = null;
const wrapper = (props = baseProps) => {
  if (!mountedWrapper) {
    mountedWrapper = shallow(<ParkingLot {...props} />);
  }
  return mountedWrapper;
};

describe('<ParkingLot />', () => {
  it('Renders the component that matches the snapshot', () => {
    const component = wrapper();
    expect(component).toMatchSnapshot();
  });
});
