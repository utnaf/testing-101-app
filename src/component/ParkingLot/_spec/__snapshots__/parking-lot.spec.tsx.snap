// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParkingLot /> Renders the component that matches the snapshot 1`] = `
<Styled(MuiBox)>
  <Component
    visible={false}
  />
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={4}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      md={4}
      onClick={[Function]}
      sm={6}
      xs={12}
    >
      <WithStyles(ForwardRef(Card))
        className="makeStyles-card-2"
      >
        <WithStyles(ForwardRef(CardMedia))
          className="makeStyles-media-3"
          title="Foo Picture"
        />
        <WithStyles(ForwardRef(CardContent))>
          <WithStyles(ForwardRef(Grid))
            container={true}
            justify="space-between"
          >
            <WithStyles(ForwardRef(Grid))
              item={true}
              xs={8}
            >
              <WithStyles(ForwardRef(Typography))
                component="h2"
                gutterBottom={true}
                variant="h5"
              >
                Foo
              </WithStyles(ForwardRef(Typography))>
            </WithStyles(ForwardRef(Grid))>
            <WithStyles(ForwardRef(Grid))
              item={true}
            >
              <WithStyles(ForwardRef(Typography))
                component="small"
                gutterBottom={true}
                variant="h5"
              >
                10
                 / 
                100
              </WithStyles(ForwardRef(Typography))>
            </WithStyles(ForwardRef(Grid))>
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Typography))>
            Bar
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(CardContent))>
      </WithStyles(ForwardRef(Card))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Fab))
    aria-label="add parking lot"
    className="makeStyles-fab-1"
    color="secondary"
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Link",
        "propTypes": Object {
          "innerRef": [Function],
          "onClick": [Function],
          "replace": [Function],
          "target": [Function],
          "to": [Function],
        },
        "render": [Function],
      }
    }
    to="/lot/create"
  >
    <AddIcon />
  </WithStyles(ForwardRef(Fab))>
</Styled(MuiBox)>
`;
